[Qt] payment request / server work - part 2

Diapolo opened the pull request:
This continues my work on the payment request / server part of our BIP70 implementation.
Please see individual commits for detailed descriptions.

Commit:
[Qt] paymentserver: do not log NULL certificates

- also add a few more comments in PaymentServer::LoadRootCAs

Commit:
[Qt] paymentserver: better logging of invalid certs

Before and after was tested in Windows:

before:
GUI: ReportInvalidCertificate : Payment server found
an invalid certificate:  ("Microsoft Authenticode(tm) Root Authority")
GUI: ReportInvalidCertificate : Payment server found
an invalid certificate:  ()
GUI: ReportInvalidCertificate : Payment server found
an invalid certificate:  ()
GUI: ReportInvalidCertificate : Payment server found
an invalid certificate:  ()

after:
GUI: ReportInvalidCertificate: Payment server found an
invalid certificate:  "01" ("Microsoft Authenticode(tm) Root Authority")
() ()
GUI: ReportInvalidCertificate: Payment server found an
invalid certificate:  "01" () () ("Copyright (c) 1997 Microsoft Corp.",
"Microsoft Time Stamping Service Root", "Microsoft Corporation")
GUI: ReportInvalidCertificate: Payment server found an
invalid certificate:  "4a:19:d2:38:8c:82:59:1c:a5:5d:73:5f:15:5d:dc:a3" ()
() ("NO LIABILITY ACCEPTED, (c)97 VeriSign, Inc.", "VeriSign Time Stamping
Service Root", "VeriSign, Inc.")
GUI: ReportInvalidCertificate: Payment server found an
invalid certificate:  "e4:9e:fd:f3:3a:e8:0e:cf:a5:11:3e:19:a4:24:02:32" ()
() ("Class 3 Public Primary Certification Authority")

Commit:
[Qt] remove unused PaymentRequestPlus::getPKIType function

Commit:
[Qt] take care of a missing typecast in PaymentRequestPlus::getMerchant()

Diapolo commented:
I'd love to see that reviewed or ACKd @laanwj @gavinandresen, mind taking a look?

Commit:
[Qt] constify first parameter of processPaymentRequest()

Commit:
[Qt] minor comment updates in PaymentServer

jonasschnelli commented on src/qt/paymentrequestplus.h:
```diff
@@ -27,7 +27,6 @@ class PaymentRequestPlus
     bool SerializeToString(std::string* output) const;
 
     bool IsInitialized() const;
-    QString getPKIType() const;
```
This method is unused. Is there a plan to make use of it soon? IMO we should keep unused implementation in master at minimum. Better keep this in a separate branch and add it if we use it.

jonasschnelli replied:
Nevermind.
You removed it. Sorry.

jonasschnelli commented:
Needs rebase.

jonasschnelli commented:
It's not related to this pull, but better here than in a new issue:
I did try creating a payment-request over gavins php tool and selected bitcoincore as merchant.
I got a `GUI: PaymentRequestPlus::getMerchant : Payment request: empty certificate chain`.
A) is there no merchant cert behind "bitcoincore"?
B) IMO we should show something to the user in case of a invalid or missing cert chain.

Diapolo commented:
@jonasschnelli Rebased

As for the cert problems with gavins php tool, IMHO you won't get any authenticated payment request from it and consensus then is to silently treat it as valid unauthenticated request. I had that discussion with @laanwj @gavinandresen and @mikehearn ;).

jonasschnelli commented:
Tested, reviewed, stepped.
Mostly move-only, cleanup things (does that require 7 commits, squash?).

ACK

Diapolo commented:
I did each change in a seperate commit to make it easy to review :).

luke-jr commented:
Does anyone have a test plan for this?

Diapolo commented:
@luke-jr Are you serious??

laanwj commented on src/qt/paymentrequestplus.cpp:
```diff
@@ -125,7 +119,7 @@ bool PaymentRequestPlus::getMerchant(X509_STORE* certStore, QString& merchant) c
     // to a valid root authority. OpenSSL needs them separately.
     STACK_OF(X509) *chain = sk_X509_new_null();
-    for (int i = certs.size()-1; i > 0; i--) {
+    for (size_t i = certs.size() - 1; i > 0; i--) {
```
NACK on changing this. There is check for certs.empty() that prevents this from becoming a bug, but for loops that count down, I generally prefer to use signed types. In this case, `i` will never visit 0, but normally a loop like this would look like:

    for (int i = certs.size() - 1; i >= 0; i--) {

If you then change the int to e.g. size_t, it will loop forever.

Diapolo replied:
Alright, will revert this change here.

Commit:
[Qt] Use identical strings for expired payment request message

- used in sendcoinsdialog.cpp and paymentserver.cpp
- removes an unneded translation string

Diapolo commented:
@jonasschnelli If you've got an idea how to group the commits so they can get merged I'll gladly do this. Guess this needs an ACK from @laanwj still :).

Diapolo commented:
@laanwj @jonasschnelli Is there some consensus possible to at least pick most of these commits or some, so we can get on here!?

Diapolo commented:
I'd like to improve payment request stuff, but not getting on with this is blocking you know ;).

laanwj commented:
ACK apart from the char* change.

Diapolo commented:
@laanwj Thanks for review, I fixed the spelling error and remove the QVariant commit, should be ready now :).
